pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -123,27 +123,48 @@ type CAOptions struct {
 	Namespace   string
 }
 
+// EnableCA returns whether CA functionality is enabled in istiod.
+// The logic of this function is from the logic of whether running CA
+// in RunCA(). The reason for moving this logic from RunCA into EnableCA() is
+// to have a central consistent endpoint to get whether CA functionality is
+// enabled in istiod. EnableCA() is called in multiple places.
+func (s *Server) EnableCA() bool {
+	if s.kubeClient == nil {
+		// No k8s - no self-signed certs.
+		// TODO: implement it using a local directory, for non-k8s env.
+		log.Warna("kubeclient is nil; disable the CA functionality", JWTPath)
+		return false
+	}
+	iss := trustedIssuer.Get()
+	if _, err := ioutil.ReadFile(JWTPath); err != nil {
+		// for debug we may want to override this by setting trustedIssuer explicitly
+		if iss == "" {
+			log.Warna("istiod running without access to K8S tokens; disable the CA functionality", JWTPath)
+			return false
+		}
+	}
+	return true
+}
+
 // RunCA will start the cert signing GRPC service on an existing server.
 // Protected by installer options: the CA will be started only if the JWT token in /var/run/secrets
 // is mounted. If it is missing - for example old versions of K8S that don't support such tokens -
 // we will not start the cert-signing server, since pods will have no way to authenticate.
-func (s *Server) RunCA(grpc *grpc.Server, opts *CAOptions, stopCh <-chan struct{}) {
-	if s.kubeClient == nil {
-		// No k8s - no self-signed certs.
-		// TODO: implement it using a local directory, for non-k8s env.
+func (s *Server) RunCA(grpc *grpc.Server, ca *ca.IstioCA, opts *CAOptions, stopCh <-chan struct{}) {
+	if !s.EnableCA() {
+		return
+	}
+	if ca == nil {
```

Comment:
Is this redundant

Reply:
No, it is not redundant: the function itself should check the input parameters.